python_modules/dagster/dagster/core/execution/context/output.py

```diff
@@ -372,6 +381,80 @@ def get_output_identifier(self) -> List[str]:
 
         return identifier
 
+    def log_event(
+        self, event: Union[AssetObservation, AssetMaterialization, Materialization]
+    ) -> None:
+        """Log an AssetMaterialization or AssetObservation from within the body of an io manager's `handle_output` method.
+
+        Events logged with this method will appear in the list of DagsterEvents, as well as the event log.
+
+        Args:
+            event (Union[AssetMaterialization, Materialization, AssetObservation]): The event to log.
+
+        Examples:
+
+        .. code-block:: python
+            from dagster import IOManager, AssetMaterialization
+
+            class MyIOManager(IOManager):
+                def handle_output(self, context, obj):
+                    context.log_event(AssetMaterialization("foo"))
+        """
+        from dagster.core.events import DagsterEvent
+
+        if isinstance(event, (AssetMaterialization, Materialization)):
+            if self._step_context:
+                self._events.append(
+                    DagsterEvent.asset_materialization(
+                        self._step_context,
+                        event,
+                        self._step_context.get_input_lineage(),
+                    )
+                )
+            self._user_events.append(event)
+        elif isinstance(event, AssetObservation):
+            if self._step_context:
+                self._events.append(DagsterEvent.asset_observation(self._step_context, event))
+            self._user_events.append(event)
+        else:
+            check.failed("Unexpected event {event}".format(event=event))
+
+    def consume_events(self) -> Iterator["DagsterEvent"]:
+        """Pops and yields all user-generated events that have been recorded from this context.
+
+        If consume_events has not yet been called, this will yield all logged events since the call to `handle_output`. If consume_events has been called, it will yield all events since the last time consume_events was called. Designed for internal use. Users should never need to invoke this method.
+        """
+
+        while self._events:
+            yield self._events.pop(0)
```

Comment:
On the op context, we discussed an approach that wouldn't require O(N^2). Does that work here?

Reply:
> What about the output context makes it so that we can't capture yielded events, unlike the op context? If it doesn't make sense to capture yielded events here, I wonder if it's better to be consistent than to capture yielded events on the op context.

@sryza 
In a testing context, `OutputContext` is just called directly with `handle_output`, there's no layer of indirection we can use to pull the yielded events into the context. Whereas with ops, we have all the machinery inside the invocation call.

> Also, we should have a context method that eliminates the need for `yield EventMetadataEntry` inside `handle_output`. Is that covered in a separate PR?

That is covered in a separate PR.

Reply:
Realizing I replied to the wrong comment, but yes that approach works here. Thanks for catching